From 664f4374f43d9c75437624f9c30f8d146b678a01 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:06:55 -0400
Subject: [ethernet] qlcnic: Optimize ring count validations

Message-id: <1408626422-3941-132-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88634
O-Subject: [PATCH RHEL 7.1 131/138] qlcnic: Optimize ring count validations
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Shahed Shaikh <shahed.shaikh@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id 18e0d625336a7690dcb358e2f4157becfa65183b

- Check interrupt mode at the start of qlcnic_set_channels().
- Do not validate ring count if they are not going to change.

Signed-off-by: Shahed Shaikh <shahed.shaikh@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
index 5bacf52..1b7f3db 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
@@ -726,6 +726,11 @@ static int qlcnic_set_channels(struct net_device *dev,
  struct qlcnic_adapter *adapter = netdev_priv(dev);
  int err;
 
+ if (!(adapter->flags & QLCNIC_MSIX_ENABLED)) {
+  netdev_err(dev, "No RSS/TSS support in non MSI-X mode\n");
+  return -EINVAL;
+ }
+
  if (channel->other_count || channel->combined_count)
   return -EINVAL;
 
@@ -734,7 +739,7 @@ static int qlcnic_set_channels(struct net_device *dev,
  if (err)
   return err;
 
- if (channel->rx_count) {
+ if (adapter->drv_sds_rings != channel->rx_count) {
   err = qlcnic_validate_rings(adapter, channel->rx_count,
          QLCNIC_RX_QUEUE);
   if (err) {
@@ -745,7 +750,7 @@ static int qlcnic_set_channels(struct net_device *dev,
   adapter->drv_rss_rings = channel->rx_count;
  }
 
- if (channel->tx_count) {
+ if (adapter->drv_tx_rings != channel->tx_count) {
   err = qlcnic_validate_rings(adapter, channel->tx_count,
          QLCNIC_TX_QUEUE);
   if (err) {
diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
index 95cfeba..5543f9f 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
@@ -4000,16 +4000,6 @@ int qlcnic_validate_rings(struct qlcnic_adapter *adapter, __u32 ring_cnt,
   strcpy(buf, "Tx");
  }
 
- if (!QLCNIC_IS_MSI_FAMILY(adapter)) {
-  netdev_err(netdev, "No RSS/TSS support in INT-x mode\n");
-  return -EINVAL;
- }
-
- if (adapter->flags & QLCNIC_MSI_ENABLED) {
-  netdev_err(netdev, "No RSS/TSS support in MSI mode\n");
-  return -EINVAL;
- }
-
  if (!is_power_of_2(ring_cnt)) {
   netdev_err(netdev, "%s rings value should be a power of 2\n",
       buf);
-- 
1.7.1